File: backend/template/tools/trade_executor.py
from .base import Tool, ToolType
from typing import Dict, Any

class TradeExecutor(Tool):
    """REACTIVE tool: Executes trades on-demand"""
    
    def _determine_type(self) -> ToolType:
        return ToolType.REACTIVE
    
    def execute(self, action: str, symbol: str, amount: float) -> Dict[str, Any]:
        """
        Executes buy/sell order
        action: "buy" or "sell"
        """
        if action == "buy":
            result = self._buy(symbol, amount)
        elif action == "sell":
            result = self._sell(symbol, amount)
        else:
            return {"status": "error", "message": "Invalid action"}
        
        return {
            "status": "success",
            "action": action,
            "symbol": symbol,
            "amount": amount,
            "result": result
        }
    
    def _buy(self, symbol: str, amount: float) -> Dict[str, Any]:
        """Platform's verified buy logic"""
        # Placeholder - would use NEAR Chain Signatures for cross-chain
        exchange = self.config.get("exchange", "near")
        return {
            "exchange": exchange,
            "order_id": f"buy_{symbol}_{amount}",
            "timestamp": "2024-01-01T00:00:00Z"
        }
    
    def _sell(self, symbol: str, amount: float) -> Dict[str, Any]:
        """Platform's verified sell logic"""
        exchange = self.config.get("exchange", "near")
        return {
            "exchange": exchange,
            "order_id": f"sell_{symbol}_{amount}",
            "timestamp": "2024-01-01T00:00:00Z"
        }
    
    def _get_config_schema(self) -> Dict[str, Any]:
        return {
            "exchange": {"type": "string", "default": "near"},
            "api_key": {"type": "string", "required": True}
        }

File: backend/utils/tools_storage/trade_executor.ts
/**
 * REACTIVE tool: Executes trades on-demand
 */
import type { ToolConfig, ExecuteResult, ConfigSchema } from "./base";

const TOOL_TYPE = "reactive" as const;

export function getConfigSchema(): ConfigSchema {
  return {
    exchange: { type: "string", default: "near" },
    api_key: { type: "string", required: true },
  };
}

function buy(symbol: string, amount: number, config: ToolConfig): ExecuteResult {
  const exchange = (config.exchange as string) ?? "near";
  return {
    status: "success",
    exchange,
    order_id: `buy_${symbol}_${amount}`,
    timestamp: new Date().toISOString(),
  };
}

function sell(symbol: string, amount: number, config: ToolConfig): ExecuteResult {
  const exchange = (config.exchange as string) ?? "near";
  return {
    status: "success",
    exchange,
    order_id: `sell_${symbol}_${amount}`,
    timestamp: new Date().toISOString(),
  };
}

/** Execute trade. Args: (config, action, symbol, amount). API passes args: ["buy"|"sell", symbol, amount]. */
export function execute(
  config: ToolConfig,
  action?: string,
  symbol?: string,
  amount?: number
): ExecuteResult {
  const act = action ?? "buy";
  const sym = symbol ?? "BTC";
  const amt = amount ?? 0;
  if (act === "buy") {
    const result = buy(sym, amt, config);
    return { status: "success", action: act, symbol: sym, amount: amt, result };
  }
  if (act === "sell") {
    const result = sell(sym, amt, config);
    return { status: "success", action: act, symbol: sym, amount: amt, result };
  }
  return { status: "error", message: "Invalid action" };
}
